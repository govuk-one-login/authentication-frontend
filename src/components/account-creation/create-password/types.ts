import type { ApiResponseResult, DefaultApiResponse } from "../../../types.js";
import type { Request } from "express";

export interface CreatePasswordServiceInterface {
  signUpUser: (
    sessionId: string,
    clientSessionId: string,
    emailAddress: string,
    password: string,
    persistentSessionId: string,
    req: Request
  ) => Promise<ApiResponseResult<DefaultApiResponse>>;
}
