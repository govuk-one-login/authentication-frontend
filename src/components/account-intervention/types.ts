import { ApiResponseResult, DefaultApiResponse } from "../../types";

export interface AccountInterventionsInterface {
  accountInterventionStatus: (
    sessionId: string,
    emailAddress: string,
    sourceIp: string,
    clientSessionId: string,
    persistentSessionId: string
  ) => Promise<ApiResponseResult<AccountInterventionStatus>>;
}

export interface AccountInterventionStatus extends DefaultApiResponse {
  email: string;
  passwordResetRequired: boolean;
  blocked: boolean;
  temporarilySuspended: boolean;
  appliedAt: string;
  reproveIdentity: boolean;
}
