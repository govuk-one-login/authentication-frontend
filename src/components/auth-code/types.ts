import {
  ApiResponseResult,
  DefaultApiResponse,
  UserSession,
  UserSessionClient,
} from "../../types";
import { Request } from "express";

export interface AuthCodeResponse extends DefaultApiResponse {
  location: string;
}

export interface AuthCodeServiceInterface {
  getAuthCode: (
    sessionId: string,
    clientSessionId: string,
    sourceIp: string,
    persistentSessionId: string,
    clientSession: UserSessionClient,
    userSession: UserSession,
    req: Request
  ) => Promise<ApiResponseResult<AuthCodeResponse>>;
}
