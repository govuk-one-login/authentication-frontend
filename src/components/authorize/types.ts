import type { MFA_METHOD_TYPE } from "../../app.constants.js";
import type { ApiResponseResult, DefaultApiResponse } from "../../types.js";
import type { Claims } from "./claims-config.js";
import type { Request } from "express";

export interface StartRequestParameters {
  authenticated: boolean;
  previous_session_id?: string;
  rp_pairwise_id_for_reauth?: string;
  previous_govuk_signin_journey_id?: string;
  reauthenticate?: string;
  cookie_consent?: string;
  _ga?: string;
  rp_client_id: string;
  scope: string;
  rp_redirect_uri: string;
  rp_state: string;
  requested_level_of_confidence?: string;
  requested_credential_strength: string;
  client_name: string;
  service_type: string;
  cookie_consent_shared: boolean;
  is_smoke_test: boolean;
  is_one_login_service: boolean;
  subject_type: string;
  is_identity_verification_required: boolean;
  rp_sector_host: string;
}

export interface StartAuthResponse extends DefaultApiResponse {
  user: UserSessionInfo;
  featureFlags?: Record<string, unknown>;
}

export interface UserSessionInfo {
  upliftRequired: boolean;
  identityRequired: boolean;
  authenticated: boolean;
  cookieConsent?: string;
  gaCrossDomainTrackingId?: string;
  mfaMethodType?: MFA_METHOD_TYPE;
  isBlockedForReauth: boolean;
}

export interface AuthorizeServiceInterface {
  start: (
    sessionId: string,
    clientSessionId: string,
    persistentSessionId: string,
    req: Request,
    startRequestParameters: StartRequestParameters
  ) => Promise<ApiResponseResult<StartAuthResponse>>;
}

export interface KmsDecryptionServiceInterface {
  decrypt(serializedJwe: string): Promise<string>;
}

export interface JwtServiceInterface {
  getPayloadWithValidation(jwt: string): Promise<Claims>;
}
