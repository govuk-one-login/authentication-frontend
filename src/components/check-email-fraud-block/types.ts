import type { ApiResponseResult, DefaultApiResponse } from "../../types.js";
import type { Request } from "express";

export interface CheckEmailFraudBlockInterface {
  checkEmailFraudBlock: (
    email: string,
    sessionId: string,
    clientSessionId: string,
    persistentSessionId: string,
    req: Request
  ) => Promise<ApiResponseResult<CheckEmailFraudBlockResponse>>;
}

export interface CheckEmailFraudBlockResponse extends DefaultApiResponse {
  email: string;
  isBlockedStatus: string;
}
