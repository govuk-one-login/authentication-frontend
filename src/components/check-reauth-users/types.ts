import { ApiResponseResult, DefaultApiResponse } from "../../types";

export interface CheckReauthServiceInterface {
  checkReauthUsers: (
    sessionId: string,
    email: string,
    sourceIp: string,
    clientSessionId: string,
    persistentSessionId: string
  ) => Promise<ApiResponseResult<DefaultApiResponse>>;
}
