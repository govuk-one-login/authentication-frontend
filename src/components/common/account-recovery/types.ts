import { ApiResponseResult, DefaultApiResponse } from "../../../types";

export interface AccountRecoveryResponse extends DefaultApiResponse {
  accountRecoveryPermitted: boolean;
}

export interface AccountRecoveryInterface {
  accountRecovery: (
    sessionId: string,
    clientSessionId: string,
    email: string,
    sourceIp: string,
    persistentSessionId: string
  ) => Promise<ApiResponseResult<AccountRecoveryResponse>>;
}
