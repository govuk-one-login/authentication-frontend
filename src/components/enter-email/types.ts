import { ApiResponseResult, DefaultApiResponse } from "../../types";

export interface UserExists extends DefaultApiResponse {
  email: string;
  doesUserExist: boolean;
  mfaMethodType: string;
  phoneNumberLastThree?: string;
}

export interface EnterEmailServiceInterface {
  userExists: (
    sessionId: string,
    emailAddress: string,
    sourceIp: string,
    clientSessionId: string,
    persistentSessionId: string
  ) => Promise<ApiResponseResult<UserExists>>;
}
