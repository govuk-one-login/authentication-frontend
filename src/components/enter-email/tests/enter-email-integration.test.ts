import request from "supertest";
import { describe } from "mocha";
import { expect, sinon } from "../../../../test/utils/test-utils";
import nock = require("nock");
import * as cheerio from "cheerio";
import decache from "decache";
import {
  API_ENDPOINTS,
  HTTP_STATUS_CODES,
  PATH_NAMES,
} from "../../../app.constants";
import { CheckReauthServiceInterface } from "../../check-reauth-users/types";
import { AxiosResponse } from "axios";
import { createApiResponse } from "../../../utils/http";
import { DefaultApiResponse } from "../../../types";

describe("Integration::enter email", () => {
  let token: string | string[];
  let cookies: string;
  let app: any;
  let baseApi: string;

  before(async () => {
    decache("../../../app");
    decache("../../../middleware/session-middleware");
    const sessionMiddleware = require("../../../middleware/session-middleware");
    const checkReauthUsersService = require("../../check-reauth-users/check-reauth-users-service");

    sinon
      .stub(sessionMiddleware, "validateSessionMiddleware")
      .callsFake(function (req: any, res: any, next: any): void {
        res.locals.sessionId = "tDy103saszhcxbQq0-mjdzU854";

        req.session.user = {
          journey: {
            nextPath: PATH_NAMES.ENTER_EMAIL_SIGN_IN,
            optionalPaths: [PATH_NAMES.SIGN_IN_OR_CREATE],
          },
          reauthenticate: "12345",
        };

        next();
      });

    sinon
      .stub(checkReauthUsersService, "checkReauthUsersService")
      .callsFake((): CheckReauthServiceInterface => {
        async function checkReauthUsers() {
          const fakeAxiosResponse: AxiosResponse = {
            status: HTTP_STATUS_CODES.OK,
          } as AxiosResponse;

          return createApiResponse<DefaultApiResponse>(fakeAxiosResponse);
        }

        return { checkReauthUsers };
      });

    app = await require("../../../app").createApp();
    baseApi = process.env.FRONTEND_API_BASE_URL;

    request(app)
      .get(PATH_NAMES.ENTER_EMAIL_SIGN_IN)
      .end((err, res) => {
        const $ = cheerio.load(res.text);
        token = $("[name=_csrf]").val();
        cookies = res.headers["set-cookie"];
      });
  });

  beforeEach(() => {
    nock.cleanAll();
  });

  after(() => {
    sinon.restore();
    app = undefined;
  });

  it("should return enter email page", (done) => {
    request(app).get(PATH_NAMES.ENTER_EMAIL_SIGN_IN).expect(200, done);
  });

  it("should return error when csrf not present", (done) => {
    request(app)
      .post(PATH_NAMES.ENTER_EMAIL_SIGN_IN)
      .type("form")
      .send({
        email: "test@test.com",
      })
      .expect(500, done);
  });

  it("should return validation error when email not entered", (done) => {
    request(app)
      .post(PATH_NAMES.ENTER_EMAIL_SIGN_IN)
      .type("form")
      .set("Cookie", cookies)
      .send({
        _csrf: token,
        email: "",
      })
      .expect(function (res) {
        const $ = cheerio.load(res.text);
        expect($("#email-error").text()).to.contains(
          "Enter your email address"
        );
      })
      .expect(400, done);
  });

  it("should return validation error when invalid email entered", (done) => {
    request(app)
      .post(PATH_NAMES.ENTER_EMAIL_SIGN_IN)
      .type("form")
      .set("Cookie", cookies)
      .send({
        _csrf: token,
        email: "test.tÂµrn@example.com",
      })
      .expect(function (res) {
        const page = cheerio.load(res.text);
        expect(page("#email-error").text()).to.contains(
          "Enter an email address in the correct format, like name@example.com\n"
        );
      })
      .expect(400);

    request(app)
      .post(PATH_NAMES.ENTER_EMAIL_SIGN_IN)
      .type("form")
      .set("Cookie", cookies)
      .send({
        _csrf: token,
        email: "test.trnexample.com",
      })
      .expect(function (res) {
        const page = cheerio.load(res.text);
        expect(page("#email-error").text()).to.contains(
          "Enter an email address in the correct format, like name@example.com\n"
        );
      })
      .expect(400);

    request(app)
      .post(PATH_NAMES.ENTER_EMAIL_SIGN_IN)
      .type("form")
      .set("Cookie", cookies)
      .send({
        _csrf: token,
        email: "test.trn@examplecom",
      })
      .expect(function (res) {
        const page = cheerio.load(res.text);
        expect(page("#email-error").text()).to.contains(
          "Enter an email address in the correct format, like name@example.com\n"
        );
      })
      .expect(400, done);
  });

  it("should redirect to /enter-password page when email address exists", (done) => {
    nock(baseApi)
      .post(API_ENDPOINTS.USER_EXISTS)
      .once()
      .reply(HTTP_STATUS_CODES.OK, {
        email: "test@test.com",
        doesUserExist: true,
      });

    request(app)
      .post(PATH_NAMES.ENTER_EMAIL_SIGN_IN)
      .type("form")
      .set("Cookie", cookies)
      .send({
        _csrf: token,
        email: "test@test.com",
      })
      .expect("Location", PATH_NAMES.ENTER_PASSWORD)
      .expect(302, done);
  });

  it("should redirect to /account-not-found when email address not found", (done) => {
    nock(baseApi).post(API_ENDPOINTS.USER_EXISTS).once().reply(200, {
      email: "test@test.com",
      doesUserExist: false,
    });

    request(app)
      .post(PATH_NAMES.ENTER_EMAIL_SIGN_IN)
      .type("form")
      .set("Cookie", cookies)
      .send({
        _csrf: token,
        email: "test@test.com",
      })
      .expect("Location", PATH_NAMES.ACCOUNT_NOT_FOUND)
      .expect(302, done);
  });

  it("should return internal server error when /user-exists API call response is 500", (done) => {
    nock(baseApi)
      .post(API_ENDPOINTS.USER_EXISTS)
      .once()
      .reply(500, {
        message: "Internal Server error",
      })
      .post("/send-notification")
      .once()
      .reply(200, {});

    request(app)
      .post(PATH_NAMES.ENTER_EMAIL_SIGN_IN)
      .type("form")
      .set("Cookie", cookies)
      .send({
        _csrf: token,
        email: "test@test.com",
      })
      .expect(500, done);
  });

  it("should redirect to /enter-password page when email address exists and check re-auth users api call is successfully", (done) => {
    process.env.SUPPORT_REAUTHENTICATION = "1";

    nock(baseApi)
      .post(API_ENDPOINTS.CHECK_REAUTH_USERS)
      .once()
      .reply(HTTP_STATUS_CODES.OK);

    nock(baseApi)
      .post(API_ENDPOINTS.USER_EXISTS)
      .once()
      .reply(HTTP_STATUS_CODES.OK, {
        email: "test@test.com",
        doesUserExist: true,
      });

    request(app)
      .post(PATH_NAMES.ENTER_EMAIL_SIGN_IN)
      .type("form")
      .set("Cookie", cookies)
      .send({
        _csrf: token,
        email: "test@test.com",
      })
      .expect("Location", PATH_NAMES.ENTER_PASSWORD)
      .expect(302, done);
  });
});
