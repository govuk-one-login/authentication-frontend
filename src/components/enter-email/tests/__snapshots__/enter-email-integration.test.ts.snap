// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration::enter email should return enter email page 1`] = `
Object {
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "sign in",
}
`;

exports[`Integration::enter email should return error when csrf not present 1`] = `
Object {
  "taxonomyLevel1": undefined,
  "taxonomyLevel2": undefined,
}
`;

exports[`Integration::enter email should return internal server error when /user-exists API call response is 500 1`] = `
Object {
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "sign in",
}
`;

exports[`Integration::enter email should return validation error when email not entered 1`] = `
Object {
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "sign in",
}
`;

exports[`Integration::enter email should return validation error when invalid email entered 1`] = `
Object {
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "sign in",
}
`;

exports[`Integration::enter email should return validation error when invalid email entered 2`] = `
Object {
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "sign in",
}
`;

exports[`Integration::enter email should return validation error when invalid email entered 3`] = `
Object {
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "sign in",
}
`;
